tools/indexscan/main.go

```diff
@@ -251,6 +252,96 @@ func (s CommitsNearPublishStrategy) Search(ctx context.Context, r *git.Repositor
 	return
 }
 
+// ExhaustiveTreeSearchStrategy
+type ExhaustiveTreeSearchStrategy struct {
+}
+
+// Search returns the set of matching commits along with the number of matches.
+func (s ExhaustiveTreeSearchStrategy) Search(ctx context.Context, r *git.Repository, hashes []string) (closest []string, matched, total int, err error) {
+	// Find subset of file hashes that are present in the repo.
+	files := make(map[plumbing.Hash]int)
+	var i int
+	for _, h := range hashes {
+		_, err = r.BlobObject(plumbing.NewHash(h))
+		if err == plumbing.ErrObjectNotFound {
+			// Skip files not present in repo.
+			err = nil
+		} else if err != nil {
+			return
+		} else {
+			files[plumbing.NewHash(h)] = i
+			i++
+		}
+	}
+	// Iterate over repo trees to create an ordering.
+	trees := make(map[plumbing.Hash]int)
+	i = 0
+	for ti, _ := r.TreeObjects(); ; {
+		t, err := ti.Next()
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			panic(err)
+		} else {
+			trees[t.Hash] = i
+			i++
+		}
+	}
+	total = len(files)
+	if total == 0 {
+		err = errors.New("repo contains no matching files")
+		return
+	}
+	// Recursively build set of matching files for each tree.
+	matches := bitmap.NewBatch(len(files), len(trees))
+	processed := bitmap.New(len(trees))
+	ti, _ := r.TreeObjects()
+	ti.ForEach(func(t *object.Tree) error {
+		processTree(t, files, trees, matches, processed)
+		return nil
+	})
+	// Search through all commits for the one whose tree has the most matches.
+	ci, _ := r.CommitObjects()
```

Comment:
check err?

Reply:
functionally, this really shouldn't happen if we've gotten this far (this is an iterator over local values) but I'll add handling